extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java

```diff
@@ -0,0 +1,238 @@
+
+public class HdfsInputSource extends AbstractInputSource implements SplittableInputSource<Path>
+{
+  private static final String PROP_PATHS = "paths";
+
+  private final List<String> inputPaths;
+  private final Configuration configuration;
+
+  @JsonCreator
+  public HdfsInputSource(
+      @JsonProperty(PROP_PATHS) Object inputPaths,
+      @JacksonInject Configuration configuration
+  )
+  {
+    // Note: Currently commas or globs in InputPaths are not supported
+    this.inputPaths = coerceInputPathsToList(inputPaths);
+
+    this.configuration = configuration;
+  }
+
+  private static List<String> coerceInputPathsToList(Object inputPaths)
+  {
+    final List<String> paths;
+
+    if (inputPaths instanceof String) {
+      paths = Collections.singletonList((String) inputPaths);
+    } else if (inputPaths instanceof List && ((List<?>) inputPaths).stream().allMatch(x -> x instanceof String)) {
+      paths = ((List<?>) inputPaths).stream().map(x -> (String) x).collect(Collectors.toList());
+    } else {
+      throw new IAE("'%s' must be a string or an array of strings", PROP_PATHS);
+    }
+
+    return paths;
+  }
+
+
+  @JsonProperty(PROP_PATHS)
+  private List<String> getInputPaths()
+  {
+    return inputPaths;
+  }
+
+  @Override
+  protected InputSourceReader formattableReader(
+      InputRowSchema inputRowSchema,
+      @Nullable InputFormat inputFormat,
+      @Nullable File temporaryDirectory
+  )
+  {
+    return new InputEntityIteratingReader(
+        inputRowSchema,
+        inputFormat,
+        createSplits(null, null).map(split -> new HdfsSource(configuration, split.get())),
+        temporaryDirectory
+    );
+  }
+
+  @Override
+  public Stream<InputSplit<Path>> createSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec)
+  {
+    return inputPaths.stream()
+                     .map(Path::new)
+                     .flatMap(path -> {
+                       try {
+                         FileSystem fs = path.getFileSystem(configuration);
+                         RemoteIterator<FileStatus> delegate = fs.listStatusIterator(fs.makeQualified(path));
+                         final Iterator<FileStatus> fileIterator = new FileStatusIterator(delegate);
+                         return StreamSupport.stream(
+                             Spliterators.spliteratorUnknownSize(
+                                 fileIterator,
+                                 Spliterator.DISTINCT & Spliterator.NONNULL
+                             ),
+                             false
+                         );
+                       }
+                       catch (IOException e) {
+                         throw new UncheckedIOException(e);
+                       }
+                     })
+                     .map(fileStatus -> new InputSplit<>(fileStatus.getPath()));
+  }
+
+  @Override
+  public int getNumSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec) throws IOException
```

Comment:
Is there an assumption baked in somewhere that `createSplits` will return a number of splits exactly equal to `getNumSplits`? If so, this implementation may violate it, since it's listing HDFS separately in each method. The underlying set of files may have changed.

Reply:
Hmm good point. Ideally, the parallel indexing task shouldn't use `numSplits` to check whether the phase is finished or not. It should just consume `Stream<InputSplit<T>>` instead.